### PR TITLE
[WIP] Quantization for generate_v2

### PR DESCRIPTION
**Pull request by joecummings**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/dev/generate_v2.py [hunk 1]
```diff
@@ -65,29 +66,37 @@ class InferenceRecipe:
...
-        - quantization through torchao
         - multi-GPU generation
         - batch generation
     """
 
     def __init__(self, cfg: DictConfig) -> None:
-        self._device = utils.get_device(device=cfg.device)
-        self._dtype = training.get_dtype(dtype=cfg.dtype, device=self._device)
-        self._logger = utils.get_logger(cfg.log_level)
+        self.device = utils.get_device(device=cfg.device)
```
Comment (joecummings) [inline recipes/dev/generate_v2.py:74]: It's a public recipe, no need to be a "private" variable. 

cc @pbontrager

File: recipes/dev/generate_v2.py [hunk 2]
```diff
@@ -65,29 +66,37 @@ class InferenceRecipe:
...
         # Instantiate model
-        with training.set_default_dtype(self._dtype), self._device:
+        with training.set_default_dtype(self.dtype), self.device:
             model = config.instantiate(cfg.model)
         model.load_state_dict(_ckpt_dict[training.MODEL_KEY])
+        self.logger.info(f"Model was initialized with precision {self.dtype}.")
+
+        # Quantize the model if specified
+        if cfg.get("quantization_method") is not None:
+            from torchao.quantization.quant_api import quantize_
```
Comment (joecummings) [inline recipes/dev/generate_v2.py:93]: Lazily import torchao API

File: recipes/dev/generate_v2.py [hunk 3]
```diff
@@ -65,29 +66,37 @@ class InferenceRecipe:
...
+
+            quantization_method = config.instantiate(cfg.quantization_method)
+            compile_model(model)
```
Comment (joecummings) [inline recipes/dev/generate_v2.py:96]: Compiling the model is necessary for quantization to be really worth it

File: recipes/dev/generate_v2.py [hunk 4]
```diff
@@ -155,20 +164,20 @@ def generate(self, cfg: DictConfig):
...
         batch["input_pos"] = input_pos[None, :seq_len]
-        utils.batch_to_device(batch, self._device)
+        utils.batch_to_device(batch, self.device)
 
         # 6. Prefill step
         generated_tokens = []
         t0 = time.perf_counter()
         logits = self.model(prompt, **batch)[:, -1]
         token = sample(logits, temperature=cfg.temperature, top_k=cfg.top_k)
+        t1 = time.perf_counter()
```
Comment (joecummings) [inline recipes/dev/generate_v2.py:179]: Now that we might have a warmup run, we log this differently so the user can see how much help quantization / compilation helps.

File: recipes/dev/generate_v2.py [hunk 5]
```diff
@@ -155,20 +164,20 @@ def generate(self, cfg: DictConfig):
                 [model_inputs], pad_direction="left", pad_max_images=1
             )
             batch["encoder_mask"] = batch["encoder_mask"][:, :seq_len]
-            prompt = batch.pop("tokens").to(self._device)
+            prompt = batch.pop("tokens").to(self.device)
         else:
-            prompt = torch.tensor(
-                model_inputs["tokens"], device=self._device
-            ).unsqueeze(0)
+            prompt = torch.tensor(model_inputs["tokens"], device=self.device)[None, :]
```
Comment (joecummings) [inline recipes/dev/generate_v2.py:169]: I wanted this to fix on one line lol

File: tests/conftest.py [hunk 6]
```diff
@@ -18,6 +19,13 @@
 CACHE_ARTIFACTS_SCRIPT_PATH = root + "/tests/cache_artifacts.sh"
 
 
+def pytest_sessionfinish():
```
Comment (joecummings) [inline tests/conftest.py:22]: Compile tries to log a bunch of stuff using the `atexit` decorator.  However, pytest closes these logs before they finish so it throws an I/O error. 

This disables logging exceptions. Not sure if the right way to do it.

File: recipes/configs/llama2/generation_v2.yaml [hunk 7]
```diff
@@ -27,16 +31,16 @@ checkpointer:
...
-dtype: bf16
-seed: 1234
-log_level: INFO
-
 # Generation arguments
 prompt:
   system: You are a helpful and creative AI assistant.
   user: What is the capital of France?
-max_new_tokens: 200
+max_new_tokens: 500
```
Comment (joecummings) [inline recipes/configs/llama2/generation_v2.yaml:38]: Allow longer generation to really see the benefit of quant + compile.

**Comment by joecummings** [comment]
@felipemello1 @ebsmothers Will this not pass on PyTorch 2.5 b/c of the issue with CUDNN? This test passes locally on PyTorch v2.5.1. 

Do we know when the patch will be released?

File: recipes/configs/llama2/generation_v2.yaml [hunk 8]
```diff
@@ -9,6 +9,10 @@
 # Model arguments
 model:
   _component_: torchtune.models.llama2.llama2_7b
+# You can turn uncomment the following lines to enable quantization for faster inference and potentially lower VRAM
+# quantization_method:
+#   _component_: torchao.quantization.quant_api.int4_weight_only # int4_weight_only is a good balance of speed and memory
```
Comment (SalmanMohammadi) [inline recipes/configs/llama2/generation_v2.yaml:14]: dumb q: so the torchtune.training.quantization API is just for QAT.. or we're not using it anymore?

File: recipes/dev/generate_v2.py (hunk 3 above)
Comment (SalmanMohammadi) [reply recipes/dev/generate_v2.py:96]: I'm curious whether compiling the model results in greater speedups than compiling the next-token-prediction fn like gptfast do

File: recipes/configs/llama2/generation_v2.yaml (hunk 8 above)
Comment (SalmanMohammadi) [reply recipes/configs/llama2/generation_v2.yaml:14]: I see you mentioned this in the PR description - if we're going to be using the torchao APIs instead it'd be good to follow up with an issue

**Comment by SalmanMohammadi** [comment]
This looks overall sensible, but a few outstanding questions I have:
- What implications does this have for how we expose quantization APIs?
- What is going on with compile?
- Why is memory usage identical for non-PTQ, and the first run of PTQ?
- Why is it so slow? Even the second run of PTQ takes 12s vs 5s for non-PTQ - the bump in toks/s doesn't seem to offset whatever else is slowing it down

**Comment by joecummings** [comment]
> * What implications does this have for how we expose quantization APIs?

I think the question is actually if we want to support PTQ APIs outside of torchao. If we do, we want want to opt for an [approach like Hugging Face's ](https://huggingface.co/docs/transformers/main/en/main_classes/quantization#quantization) wherein a config for a specific backend can be initialized. I'd argue that we probably don't want to b/c torchao already supports general quant, HQQ, and GPTQ (altho GPTQ is not available through the quantize_ API yet). Idk if this is too short sighted though.

> * What is going on with compile?

Not sure I understand the question. It's always slow during warmup run. 

> * Why is memory usage identical for non-PTQ, and PTQ? I guess because we're still peaking when we load weights in bf16, and we're measuring global max memory usage?

Exactly.

> * Why is it so slow? Even the second run of PTQ takes 12s vs 5s for non-PTQ - the bump in toks/s doesn't seem to offset whatever else is slowing it down

Not sure *what* is so slow, but I've reached out to the AO team to see if this is normal.

> * Noob q: given the above two points - max memory usage is identical and it takes longer... when would someone want to use this?

An excellent question. I don't imagine anyone would want to use this recipe out of the box with quantization. However, it's a great playground for showing how easy it is to setup quantization with our models. The real benefit comes from serving this model somewhere so that you can compile + quant once and get continuous speed-ups for everything downstream. Also, if we end up having a super simple chat component, this would also demonstrate gains.

**Comment by andrewor14** [comment]
Are you seeing the slowdown for `int4_weight_only` specifically? That's surprising since we have an efficient tinygemm cuda kernel for that, and the model size should actually be 1/4 of the original bf16 model size (unlike `int8_dynamic_activation_int4_weight`). Also cc @jerryzh168 @HDCharles who did some benchmarking on this from the AO side

File: recipes/configs/llama2/generation_v2.yaml [hunk 9]
```diff
@@ -9,6 +9,10 @@
...
+#   use_hqq: False # Turn on for more accurate results
```
Comment (andrewor14) [inline recipes/configs/llama2/generation_v2.yaml:15]: @HDCharles is this true?

Comment (joecummings) [reply recipes/configs/llama2/generation_v2.yaml:15]: Yeah sorry this was anecdotal.

**Comment by joecummings** [comment]
> Are you seeing the slowdown for `int4_weight_only` specifically? 

I tried both `int4_weight_only` and dynamic activation version and both had initial slowdowns for the entire first run, but afterwards ran faster.

File: recipes/dev/generate_v2.py (hunk 3 above)
Comment (jerryzh168) [reply recipes/dev/generate_v2.py:96]: we should compile after `quantize_` for speedup actually

Comment (joecummings) [reply recipes/dev/generate_v2.py:96]: Interesting! I was following the pattern from AO's README where the model is compiled first:

```python
model = torchao.autoquant(torch.compile(model, mode='max-autotune'))
```

Why should the model be compiled after quantization?

**Comment by joecummings** [comment]
> > > Are you seeing the slowdown for `int4_weight_only` specifically?
> > 
> > 
> > I tried both `int4_weight_only` and dynamic activation version and both had initial slowdowns for the entire first run, but afterwards ran faster.
> 
> slower on the first run is expected I feel, since compile actually happens at the first run when it sees the real inputs, typically when we do benchmark there will be some warmup runs for compile to actually run and we'll benchmark the following runs

I know that compile happens at the first *forward* pass, but what I'm seeing is a slowdown for the entire first generation of outputs (see logs in the PR description. Is this expected?

File: recipes/dev/generate_v2.py (hunk 3 above)
Comment (joecummings) [reply recipes/dev/generate_v2.py:96]: @jerryzh168 Anecdotally, I don't see much difference in tok/sec (after first token) between putting compile first or second. Can you share some more details about which one is correct?

File: recipes/configs/llama2/generation_v2.yaml [hunk 10]
```diff
@@ -9,6 +9,10 @@
 # Model arguments
 model:
   _component_: torchtune.models.llama2.llama2_7b
+# You can turn uncomment the following lines to enable quantization for faster inference and potentially lower VRAM
+# quantization_method:
+#   _component_: torchao.quantization.quant_api.int4_weight_only # int4_weight_only is a good balance of speed and memory
+#   use_hqq: False # Turn on to use Half-Quadratic Quantization
```
Comment (felipemello1) [inline recipes/configs/llama2/generation_v2.yaml:15]: what does it mean? Can you add if it makes it faster/more accurate/less memory?

Comment (felipemello1) [reply recipes/configs/llama2/generation_v2.yaml:15]: sorry, what i meant is that this should be made clear for the user in the comment :P

File: recipes/dev/generate_v2.py (hunk 3 above)
Comment (jerryzh168) [reply recipes/dev/generate_v2.py:96]: oh right now `quantize_` needs to compile after: https://github.com/pytorch/ao/blob/main/torchao/quantization/README.md#full-affine-quantization-flow-example

but `autoquant` will do compile first before calling `autoquant`

Comment (felipemello1) [reply recipes/dev/generate_v2.py:96]: mode='max-autotune' will take a long time to compile. Is it worth it? We dont do it for training.

Its interesting that in AO's read it says to put compile first. Do we also do it for QLoRA?

Comment (jerryzh168) [reply recipes/dev/generate_v2.py:96]: > I don't see much difference in tok/sec (after first token) between putting compile first or second.

I haven't tried calling `quantize_` after compile actually, maybe it would have the same effect as well, need to confirm

Comment (jerryzh168) [reply recipes/dev/generate_v2.py]: > Its interesting that in AO's read it says to put compile first. Do we also do it for QLoRA?

This is only for autoquant actually, we also haven't tested QLoRA, I think @andrewor14 is taking a look now

If compile time is a concern, we are also thinking of just do autoquant before hand and save the model, but I'm still testing that path as well
